File: tui/src/scala/tui/StatefulWidget.scala
package tui

/// A `StatefulWidget` is a widget that can take advantage of some local state to remember things
/// between two draw calls.
///
/// Most widgets can be drawn directly based on the input parameters. However, some features may
/// require some kind of associated state to be implemented.
///
/// For example, the [`List`] widget can highlight the item currently selected. This can be
/// translated in an offset, which is the number of elements to skip in order to have the selected
/// item within the viewport currently allocated to this widget. The widget can therefore only
/// provide the following behavior: whenever the selected item is out of the viewport scroll to a
/// predefined position (making the selected item the last viewable item or the one in the middle
/// for example). Nonetheless, if the widget has access to the last computed offset then it can
/// implement a natural scrolling experience where the last offset is reused until the selected
/// item is out of the viewport.
///
/// ## Examples
///
/// ```rust,no_run
/// # use std::io;
/// # use tui::Terminal;
/// # use tui::backend::{Backend, TestBackend};
/// # use tui::widgets::{Widget, List, ListItem, ListState};
///
/// // Let's say we have some events to display.
/// struct Events {
///     // `items` is the state managed by your application.
///     items: Vec<String>,
///     // `state` is the state that can be modified by the UI. It stores the index of the selected
///     // item as well as the offset computed during the previous draw call (used to implement
///     // natural scrolling).
///     state: ListState
/// }
///
/// impl Events {
///     fn new(items: Vec<String>) -> Events {
///         Events {
///             items,
///             state: ListState::default(),
///         }
///     }
///
///     pub fn set_items(&mut self, items: Vec<String>) {
///         self.items = items;
///         // We reset the state as the associated items have changed. This effectively reset
///         // the selection as well as the stored offset.
///         self.state = ListState::default();
///     }
///
///     // Select the next item. This will not be reflected until the widget is drawn in the
///     // `Terminal::draw` callback using `Frame::render_stateful_widget`.
///     pub fn next(&mut self) {
///         let i = match self.state.selected() {
///             Some(i) => {
///                 if i >= self.items.len() - 1 {
///                     0
///                 } else {
///                     i + 1
///                 }
///             }
///             None => 0,
///         };
///         self.state.select(Some(i));
///     }
///
///     // Select the previous item. This will not be reflected until the widget is drawn in the
///     // `Terminal::draw` callback using `Frame::render_stateful_widget`.
///     pub fn previous(&mut self) {
///         let i = match self.state.selected() {
///             Some(i) => {
///                 if i == 0 {
///                     self.items.len() - 1
///                 } else {
///                     i - 1
///                 }
///             }
///             None => 0,
///         };
///         self.state.select(Some(i));
///     }
///
///     // Unselect the currently selected item if any. The implementation of `ListState` makes
///     // sure that the stored offset is also reset.
///     pub fn unselect(&mut self) {
///         self.state.select(None);
///     }
/// }
///
/// # let backend = TestBackend::new(5, 5);
/// # let mut terminal = Terminal::new(backend).unwrap();
///
/// let mut events = Events::new(vec![
///     String::from("Item 1"),
///     String::from("Item 2")
/// ]);
///
/// loop {
///     terminal.draw(|f| {
///         // The items managed by the application are transformed to something
///         // that is understood by tui.
///         let items: Vec<ListItem>= events.items.iter().map(|i| ListItem::new(i.as_ref())).collect();
///         // The `List` widget is then built with those items.
///         let list = List::new(items);
///         // Finally the widget is rendered using the associated state. `events.state` is
///         // effectively the only thing that we will "remember" from this draw call.
///         f.render_stateful_widget(list, f.size(), &mut events.state);
///     });
///
///     // In response to some input events or an external http request or whatever:
///     events.next();
/// }
/// ```
trait StatefulWidget {
  type State;
  def render(area: Rect, buf: Buffer, state: State): Unit
}

File: jatatui-core/src/java/jatatui/core/widgets/StatefulWidget.java
package jatatui.core.widgets;

import jatatui.core.buffer.Buffer;
import jatatui.core.layout.Rect;

/// A `StatefulWidget` is a widget that can take advantage of some local state to remember things
/// between two draw calls.
///
/// Mirrors `ratatui_core::widgets::StatefulWidget` (v0.30).
///
/// Most widgets can be drawn directly based on the input parameters. However, some features
/// require associated state to be implemented.
///
/// For example, the `List` widget can highlight the item currently selected. This is translated
/// to an offset (number of elements to skip so the selected item is within the viewport allocated
/// to the widget). Without state, the widget can only scroll to a predefined position whenever
/// the selected item leaves the viewport. With state — the last computed offset stored across
/// draws — the widget can implement natural scrolling: the last offset is reused until the
/// selected item leaves the viewport.
///
/// In Rust this is a trait with an associated type `State`. In Java we model the state as a
/// regular type parameter on the interface.
///
/// **`StatefulWidgetRef` — N/A**: the upstream documentation mentions a historical
/// `StatefulWidgetRef` trait, but it does not exist in `ratatui-core` v0.30.
///
/// **Unsized state types** (`type State = [u8]`): Rust permits `?Sized` associated types so the
/// state can be a slice. Java has no unsized types; pass the array (or any other reference) as
/// the type parameter and mutate it in place.
///
/// # Example
///
/// ```java
/// class PersonalGreeting implements StatefulWidget<StringBuilder> {
///   public void render(Rect area, Buffer buf, StringBuilder state) {
///     Widget.renderString("Hello " + state, area, buf);
///   }
/// }
/// ```
@FunctionalInterface
public interface StatefulWidget<State> {

  /// Draws the current state of the widget in the given buffer. That is the only method required
  /// to implement a custom stateful widget.
  ///
  /// Upstream signature is `fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State)`
  /// — `self` is consumed and `state` is mutably borrowed. Java has no consume semantics; the
  /// widget should be treated as logically consumed after this call. The `state` parameter is a
  /// regular reference and is expected to be mutated in place.
  void render(Rect area, Buffer buf, State state);
}
